includes/admin/extensions/abstract-extension.php

```diff
@@ -0,0 +1,312 @@
+			$this->get_configuration( $product_data )
+		);
+	}
+
+	/**
+	 * Gets the product data for a specific extension.
+	 *
+	 * @param false|int $item_id
+	 * @return bool|\ProductData|array False if there is no data; product data object if there is, or possibly an array of arrays.
+	 */
+	public function get_product_data( $item_id = false ) {
+		require_once EDD_PLUGIN_DIR . 'includes/admin/extensions/class-extensions-api.php';
+		$api          = new ExtensionsAPI();
+		$body         = $this->get_api_body();
+		$api_item_id  = $item_id ? $item_id : $this->item_id;
+		$product_data = $api->get_product_data( $body, $api_item_id );
+		if ( ! $product_data ) {
+			return false;
+		}
+
+		if ( $api_item_id ) {
+			return $product_data;
+		}
+
+		if ( $item_id && ! empty( $product_data[ $item_id ] ) ) {
+			return $product_data[ $item_id ];
+		}
+
+		return $product_data;
+	}
+
+	/**
+	 * Gets the custom configuration for the extension.
+	 *
+	 * @since 2.11.x
+	 * @param \EDD\Admin\Extensions\ProductData $product_data Optionally allows the product data to be parsed inthe configuration.
+	 * @return array
+	 */
+	protected function get_configuration( \EDD\Admin\Extensions\ProductData $product_data ) {
+		return array();
+	}
+
+	/**
+	 * Whether the current screen is an EDD setings screen.
+	 *
+	 * @since 2.11.x
+	 * @return bool
+	 */
+	protected function is_edd_settings_screen() {
+		return edd_is_admin_page( 'settings', $this->settings_tab );
+	}
+
+	/**
+	 * Whether the current screen is a download new/edit screen.
+	 *
+	 * @since 2.11.x
+	 * @return bool
+	 */
+	protected function is_download_edit_screen() {
+		return edd_is_admin_page( 'download', 'edit' ) || edd_is_admin_page( 'download', 'new' );
+	}
+
+	/**
+	 * Gets the array for the body of the API request.
+	 * Classes may need to override this (for example, to query a specific tag).
+	 * Note that the first array key/value pair are used to create the option name.
+	 *
+	 * @return array
+	 */
+	protected function get_api_body() {
+		return array();
+	}
+
+	/**
+	 * Gets the type for the button data-type attribute.
+	 * This is intended to sync with the Products API request.
+	 * Default is product.
+	 *
+	 * Really a shim for array_key_first.
+	 *
+	 * @param array $array
+	 * @return string
+	 */
+	private function get_type( array $array ) {
+		$type = 'product';
+		if ( empty( $array ) ) {
+			return $type;
+		}
+		if ( function_exists( 'array_key_first' ) ) {
+			return array_key_first( $array );
+		}
+		foreach ( $array as $key => $unused ) {
+			return $key;
+		}
+
+		return $type;
+	}
+
+	/**
+	 * Gets the button parameters.
+	 * Classes should not need to replace this method.
+	 *
+	 * @param \EDD\Admin\Extensions\ProductData $product_data The extension data returned from the Products API.
+	 * @param int|false                         $item_id      Optional: the item ID.
+	 * @return array
+	 */
+	protected function get_button_parameters( $product_data, $item_id = false ) {
```

Comment:
```suggestion
	protected function get_button_parameters( ProductData $product_data, $item_id = false ) {
```

We're actually allowed to type hint classes! (Assuming it has to be that class, which it looks like it does.)